src/main/scala/firrtl/passes/NewLowerTypes.scala

```diff
@@ -0,0 +1,439 @@
+    // replace all declarations and references with the destructed types
+    implicit val symbols: LoweringTable = new LoweringTable(scan, renameMap, ref)
+    implicit val memInit: Seq[MemoryInitAnnotation] = memoryInit
+
+    val loweredPorts = m.ports.flatMap(symbols.lower) // we cannot use mapPort as ports might be expanded
+    val newMod = m match {
+      case e : ExtModule => e.copy(ports = loweredPorts)
+      case mod: Module => mod.copy(ports = loweredPorts).mapStmt(onStatement)
+    }
+
+    (newMod, renameMap, memInit)
+  }
+
+  def onPort(p: Port)(symbols: LoweringTable): Seq[Port] = symbols.lower(p)
+
+  def onStatement(s: Statement)(implicit symbols: LoweringTable, memInit: Seq[MemoryInitAnnotation]): Statement = s match {
+    // declarations
+    case d : DefWire =>
+      Block(symbols.lower(d.name, d.tpe, firrtl.WireKind).map{case (name, tpe, _) => d.copy(name=name, tpe=tpe) })
+    case d @ DefRegister(info, _, _, clock, reset, _) =>
+      // clock and reset are always of ground type
+      val loweredClock = onExpression(clock)
+      val loweredReset = onExpression(reset)
+      // It is important to first lower the declaration, because the reset can refer to the register itself!
+      val loweredRegs = symbols.lower(d.name, d.tpe, firrtl.RegKind)
+      val inits = Utils.create_exps(d.init).map(onExpression)
+      Block(
+        loweredRegs.zip(inits).map { case ((name, tpe, _), init) =>
+          DefRegister(info, name, tpe, loweredClock, loweredReset, init)
+      })
+    case d : DefNode =>
+      val values = Utils.create_exps(d.value).map(onExpression)
+      Block(
+        symbols.lower(d.name, d.value.tpe, firrtl.NodeKind).zip(values).map{ case((name, tpe, _), value) =>
+          assert(tpe == value.tpe)
+          DefNode(d.info, name, value)
+      })
+    case d : DefMemory =>
+      // TODO: as an optimization, we could just skip ground type memories here.
+      //       This would require that we don't error in getReferences() but instead return the old reference.
+      val mems = symbols.lower(d)
+      if(mems.length > 1 && memInit.exists(_.target.ref == d.name)) {
+        val mod = memInit.find(_.target.ref == d.name).get.target.encapsulatingModule
+        val msg = s"[module $mod] Cannot initialize memory ${d.name} of non ground type ${d.dataType.serialize}"
+        throw new RuntimeException(msg)
+      }
+      Block(symbols.lower(d))
+    case d : DefInstance => symbols.lower(d)
+    // connections
+    case Connect(info, loc, expr) =>
+      if(!expr.tpe.isInstanceOf[GroundType]) {
+        throw new RuntimeException(s"NewLowerTypes expects Connects to have been expanded! ${expr.tpe.serialize}")
+      }
+      val rhs = onExpression(expr)
+      // We can get multiple refs on the lhs because of ground-type memory ports like "clk" which can get duplicated.
+      val lhs = symbols.getReferences(loc.asInstanceOf[RefLikeExpression])
```

Comment:
I wonder if it would be useful to change the `loc` type in `Connect` to be of `RefLikeExpression`. A future PR perhaps, this is fine as is.

Reply:
I was trying to do that, but it messes with a lot of old code and requires you to insert quite some `asInstanceOf[]` casts. So yeah, maybe in the future.